### PR TITLE
Pile, Stack, Query, and GermlineGeneMap.

### LINKED ISSUE #24: build an object to contain information from the partis CSV
Member variables:

* name
* sequence
* flexbounds
* relpos

Be able to parse from CSV

### PR DESCRIPTION
**Pull request by matsen**
Closes #24, #5, #13, #25, #26, #7.

## Individual Commits
template for a Germline Gene map

changing data directory for testing purposes

some GermlineGene map tests

Implementing a Query class for partis queries.

Closes #24.

removing unnecessary YAML function

fixed build error

added VDJ pointer cast member functions

slight test changes

slight Query const'ing

added alphabet map to Germline class

fixed alphabet map function in Germline

more needed Query class changes

changed VDJSmooshable functions to adjust to Query design

Initial implementation of new SmooshableChain.

Why are my matrices getting corrupted?

Corruption fixed thanks to @bcclaywell .

Initial implementation of SmooshableChain.

Not compiling in clang.

Moving to an operator[] for SmooshableStack.

Adding a push_back to SmooshableStack.

Viterbi WIP.

Adding "Smooshish" abstract base class.

Removing Smooshable constructor by size.

Member variables private for Smooshable and Chain.

Getting Viterbi paths with new SmooshableChain.

Changing around signature of AuxViterbiPath.

FinalViterbiLogProb; untested.

Extending laziness by storing flexes.

Factoring out the scaling in SmooshableChain.

Adding an iterator to SmooshableStack.

Smooshish now has the full set of methods.

@bcclaywell: I have a situation where there are two classes that share
most of an interface: Smooshables implement a subset of the
functionality of SmooshableChains. In this commit I made their abstract
base class require all of the functionality of the SmooshableChain. We
want to be able to store either in SmooshableStacks, which seems to
imply that we would either need to do some dynamic casting, or something
like this.

Comments?

FinalViterbiLogProb completion and test.

Adding a SmooshableChain underflow test.

SmooshableStack -> Pile; SmooshableChain -> Chain.

SmooshVector chains multiple Smooshables on right.

AuxViterbiPath documentation.

DJSmooshable now maintain NTI information separate

fixed tests + added new Pile Smoosher

Adding `reserve` to Pile.

Updating variable names in tests.

Docs tweaks in Chain.

Adding in my ycm_extra_conf.

extra Query assert

Pile Maker function + NO TESTS

added ConvertSeqToInts function declaration in header file

A few comments and edits.

added comment to GermlineGeneMap function

reshaping data directory

more data changes

changing example YAML names

added example partis CSV file

beginning of VDJ pile tests

slight naming change

temporary tests

necessary changes to GermlineProbMatrix

tightening assert's for main probability calculators

another assert change

re-jiggering Smooshable computation - partially done

more re-jiggering of Smooshables

added proper VDJ pile tests

made more detailed Germline Smooshable tests

A comment in Chain's SmooshVector.

test.cpp tidying.

### CODE REVIEW

File: data/hmm_params_real/IGHJ4_star_01.yaml [hunk 1]
```diff
@@ -437,6 +437,7 @@ states:
     track: nukes
   extras: {ambiguous_char: N, ambiguous_emission_prob: 0.25, germline: N}
   name: insert_right_N
-  transitions: {end: 0.040000000000000036, insert_right_N: 0.96}
+  # Note that the "end" transition probability should be 0.040000000000000036.
```
Comment (dunleavy005) [inline data/hmm_params_real/IGHJ4_star_01.yaml]: Duncan's J gene partis files have an `end` probability of 0.040000000000000036 in the `insert_right_N` state, BUT have an `end` probability of 0.04 in the last germline state before it.  Technically, the `end` probabilities should be identical and we have an assertion that checks this (https://github.com/matsengrp/linearham/blob/25-gname-pr/src/NPadding.cpp#L62). I modified our example J gene file just to get the code running, and commented that I made the modification.

File: src/Chain.cpp [hunk 2]
```diff
@@ -0,0 +1,136 @@
...
+  prev_->AuxViterbiPath(row, new_col, path);
+  path.push_back(new_col);
+};
+
+
+/// @brief Give the Viterbi path corresponding to the given row and column of
+/// the Viterbi matrix.
+std::vector<int> Chain::ViterbiPath(int row, int col) const {
+  std::vector<int> path;
+  /// @todo increase to 5?
```
Comment (dunleavy005) [inline src/Chain.cpp]: Done.

File: data/hmm_params_real/IGHJ4_star_01.yaml (hunk 1 above)
Comment (matsen) [inline data/hmm_params_real/IGHJ4_star_01.yaml]: Got it. Would be great to just throw your first sentence in the comment. Thanks!

Comment (dunleavy005) [inline data/hmm_params_real/IGHJ4_star_01.yaml]: Done.

File: src/Smooshable.cpp [hunk 3]
```diff
@@ -6,216 +6,245 @@
...
-      V_left_flexbounds, emission_indices, V_relpos, true);
-  germ_prob_row *= npadding_prob;
+      flexbounds.at("v_l"), emission_indices, v_relpos, true);
+  germ_prob_matrix *= npadding_prob;
 
-  // storing Smooshable object
-  Smooshable v_smooshable = Smooshable(germ_prob_row);
-
-  return v_smooshable;
+  /// @todo do we want some move semantics here?
```
Comment (bcclaywell) [inline src/Smooshable.cpp]: @matsen @dunleavy005 I agree with "no", it looks like this and `BuildSmooshablePtr` are doing the right thing.

File: src/Germline.hpp [hunk 4]
```diff
@@ -13,10 +13,12 @@ namespace linearham {
 /// reads.
 class Germline {
  protected:
-  Eigen::VectorXd landing_;
+  Eigen::VectorXd landing_in_;
+  Eigen::VectorXd landing_out_;
```
Comment (dunleavy005) [inline src/Germline.hpp]: Done.

File: src/Smooshable.cpp [hunk 5]
```diff
@@ -6,216 +6,245 @@
...
-      vgerm_obj
-          .GermlineProbMatrix(V_left_flexbounds, V_right_flexbounds,
-                              emission_indices, V_relpos)
-          .row(start_pos);
+/// A SmooshablePtr.
+SmooshablePtr VSmooshable(
+    const VGermline& vgerm_obj,
+    const std::map<std::string, std::pair<int, int>>& flexbounds,
+    const Eigen::Ref<const Eigen::VectorXi>& emission_indices, int v_relpos) {
+  // computing the germline match probability matrix
```
Comment (dunleavy005) [inline src/Smooshable.cpp]: Done, changed all gerund phrases in ALL files in linearham.

File: src/Smooshable.cpp [hunk 6]
```diff
@@ -6,216 +6,245 @@
...
 
-/// @brief Multiplies a landing vector into a germline match probability
-/// matrix.
+/// @brief Multiplies a landing probability vector into a germline match
+/// probability matrix.
 /// @param[in] landing
-/// A vector of probabilities of landing somewhere to begin the germline
+/// A vector of landing probabilities to either begin or end the germline
 /// match.
 /// @param[in,out] germ_prob_matrix
```
Comment (dunleavy005) [inline src/Smooshable.cpp]: Done.

**Pull request merged**